client/src/randomBag.ts

```diff
@@ -0,0 +1,34 @@
+import { TetrominoType } from "common/TetrominoType";
+
+function fisherShuffle(arr: Array<any>) {
+    let currIdx = arr.length - 1;
+
+    while (currIdx > 0) {
+        const randIdx = Math.floor(Math.random() * currIdx);
+        [arr[currIdx], arr[randIdx]] = [arr[randIdx], arr[currIdx]];
+        currIdx--;
+    }
+    return arr;
+}
+export class RandomBag{
+    bagStack: Array<TetrominoType>;
+
+    constructor() {
+        this.bagStack = [];
+
+    }
+    private createBag() {
+        const bagList = [TetrominoType.I, TetrominoType.J, TetrominoType.L, TetrominoType.O, TetrominoType.S, TetrominoType.T, TetrominoType.Z];
+        fisherShuffle(bagList);
+        this.bagStack = bagList;
+    }
+    public returnNextPiece() {
+        if (this.bagStack.length == 0) {
+            this.createBag();
+        }
+        const nextPiece: TetrominoType = this.bagStack[this.bagStack.length - 1];
```

Comment:
could this line be replaced with `const nextPiece: TetrominoType = this.bagStack.pop();`?
If so you could get rid of next line

Reply:
The problem is that the return type for pop is "[TYPE] | UNDEFINED". And it seemed like typescript couldn't infer that the array wouldn't be non-empty when pop is called. So using pop just results in type errors since then it thinks the type has a possibility of being undefined.

Reply:
Ah, I see, makes sense 👍